tests/backends/postgresql/test_server_side_cursors.py

```diff
@@ -31,17 +31,20 @@ def inspect_cursors(self):
 
     @contextmanager
     def override_db_setting(self, **kwargs):
-        for setting in kwargs:
-            original_value = connection.settings_dict.get(setting)
-            if setting in connection.settings_dict:
-                self.addCleanup(
-                    operator.setitem, connection.settings_dict, setting, original_value
-                )
+        cleanups = []
+        settings = connection.settings_dict
+
+        for name, value in kwargs.items():
+            if name in settings:
+                cleanups.append((operator.setitem, settings, name, settings[name]))
             else:
-                self.addCleanup(operator.delitem, connection.settings_dict, setting)
+                cleanups.append((operator.delitem, settings, name))
+            settings[name] = value
+
+        yield
 
-            connection.settings_dict[setting] = kwargs[setting]
-            yield
+        for operation, *args in reversed(cleanups):
+            operation(*args)
```

Comment:
I don't see how this change is related with PyPy :thinking:

Reply:
This was something I looked at while trying to see if I could get the `test_server_side_cursors_setting` working. Things were behaving strangely and I saw that this context manager did things in a strange way as detailed in the message for that commit:

> The `yield` should be after all settings changes have been applied (not that this context manager is used with more than one setting override) and the clean up should happen when the context manager exits, not after the end of the test. This is because the context manager is used multiple times within the same test.

Ultimately this didn't help with the issue I had around cursor destruction - so it's not related - but I left the commit because it seemed more correct. If you don't think that makes sense, feel free to drop it.

Reply:
Let's move this to a separate PR.

Reply:
Extracted to #17578.